system/mp/mpi_wrapper.cpp

```diff
@@ -0,0 +1,519 @@
+    {
+        if (asyncCommData[i]==commData)
+        {
+            _popCommData(i);
+            break;
+        }
+    }
+    return commData;
+}
+
+//----------------------------------------------------------------------------------------------------------------------------//
+
+#define TAG_UB 1000000 //INT_MAX //MPI_TAG_UB
+int getRank(rank_t sourceRank)
+{
+    _TF("getRank", sourceRank);
+    if (sourceRank == RANK_ALL)
+        return MPI_ANY_SOURCE;
+    else
+        return sourceRank;
+}
+
+#define SPECIAL_TAG_BASE (TAG_UB/2) // reserve half range for special tags (inc. reply tags)
+int getMPITag(mptag_t mptag)
+{
+    _TF("getMPITag", mptag);
+
+    if (mptag == TAG_ALL)
+        return MPI_ANY_TAG;
+    else
+    {
+        int tag = (int)mptag;
+        if (tag < 0)
+        {
+            assertex(-tag < TAG_UB-SPECIAL_TAG_BASE); // check not out of range
+            return -tag + SPECIAL_TAG_BASE;
+        }
+        else
+        {
+            assertex(tag <= TAG_UB);
+            return tag;
+        }
+    }
+}
+
+mptag_t getMPTag(int tag)
+{
+    _TF("getMPTag", tag);
+    unsigned returnVal;
+    if (tag == MPI_ANY_TAG)
+        returnVal = TAG_ALL;
+    else
+    {
+        if (tag > SPECIAL_TAG_BASE)
+            returnVal = (mptag_t) -(tag-SPECIAL_TAG_BASE);
+        else
+            returnVal = tag;
+    }
+    return (mptag_t)returnVal;
+}
+
+MPI_Status waitToComplete( bool& completed, bool& error, bool& canceled, bool& timedout, unsigned timeout, CommData *commData)
+{
+    _TF("waitToComplete", completed, error, canceled, timeout);
+    CTimeMon tm(timeout);
+    MPI_Status stat;
+    unsigned remaining;
+    bool noCancellation;
+    int flag;
+    while (!(completed || error || (timedout = tm.timedout(&remaining))) && (noCancellation = commData->lockFromCancellation()))
+    {
+
+        MPI_Test(commData->request(), &flag, &stat);
+        completed = flag > 0;
+        commData->releaseCancellationLock();
+        usleep(WAIT_DELAY);
+    }
+    if (completed)
+    {
+        MPI_Test_cancelled( &stat, &flag);
+        canceled = flag > 0;
+    }
+    else
+        canceled = !noCancellation;
+
+    return stat;
+}
+
+MPI_Status hasIncomingData(int sourceRank, int mptag, hpcc_mpi::MPIComm& comm,
+        bool& incomingMessage, bool& error, bool& canceled, unsigned timeout, CommData *commData)
+{
+    _TF("hasIncomingData", sourceRank, mptag, incomingMessage, error, timeout);
+
+    CTimeMon tm(timeout);
+    MPI_Status stat;
+
+    int flag;
+    unsigned remaining;
+    bool noCancellation;
+    while (!(incomingMessage || error || (timeout !=0 && tm.timedout(&remaining))) && (noCancellation = commData->lockFromCancellation()))
+    {
+        MPI_Iprobe(sourceRank, mptag, comm.get(), &flag, &stat);
+        incomingMessage = flag > 0;
+        commData->releaseCancellationLock();
+        if (timeout == 0) break;
+        usleep(PROBING_DELAY);
+    }
+    canceled = !noCancellation;
```

Comment:
noCancellation could be uninitialized here..

Reply:
fixed in both places.